File: examples/native-capture/guardian_openai.mjs
// OpenAI usage adapters. Import is inert; the application owns its SDK and requests.
import { types } from 'node:util';

const MAX_TOKENS = Number.MAX_SAFE_INTEGER;
const UNKNOWN = () => ({ input_tokens: null, output_tokens: null, total_tokens: null });
const APIS = new Set(['responses', 'chat_completions']);
const SUCCESS = new Set(['stop', 'tool_calls', 'function_call']);
const TERMINAL = new Set([...SUCCESS, 'length', 'content_filter']);
const RESPONSE_TERMINAL = new Set(['completed', 'failed', 'incomplete']);

function record(value) {
  return value !== null && typeof value === 'object' && !types.isProxy(value)
    && !Array.isArray(value) && [null, Object.prototype].includes(Object.getPrototypeOf(value));
}
function field(value, name) {
  if (!record(value)) throw new Error('unsupported_usage');
  const descriptor = Object.getOwnPropertyDescriptor(value, name);
  if (!descriptor) return undefined;
  if (!Object.hasOwn(descriptor, 'value')) throw new Error('unsupported_usage');
  return descriptor.value;
}
function array(value, max = 1) {
  if (!Array.isArray(value) || types.isProxy(value)) throw new Error('unsupported_usage');
  const length = Object.getOwnPropertyDescriptor(value, 'length').value;
  if (length > max) throw new Error('unsupported_usage');
  const items = [];
  for (let index = 0; index < length; index += 1) {
    const item = Object.getOwnPropertyDescriptor(value, String(index));
    if (!item || !Object.hasOwn(item, 'value')) throw new Error('unsupported_usage');
    items.push(item.value);
  }
  return items;
}
function count(value) {
  if (value === undefined || value === null) return null;
  if (!Number.isSafeInteger(value) || value < 0 || value > MAX_TOKENS) throw new Error('unsupported_usage');
  return value;
}
function subset(usage, detailsName, names, parent, total) {
  const details = field(usage, detailsName);
  if (details === undefined || details === null) return;
  for (const name of names) {
    const value = count(field(details, name));
    if (value !== null && [parent, total].some((limit) => limit !== null && value > limit)) throw new Error('unsupported_usage');
  }
}
function usageSnapshot(response, api) {
  if (!APIS.has(api)) throw new Error('unsupported_api');
  const usage = field(response, 'usage');
  if (usage === undefined || usage === null) return UNKNOWN();
  const input = count(field(usage, api === 'responses' ? 'input_tokens' : 'prompt_tokens'));
  const output = count(field(usage, api === 'responses' ? 'output_tokens' : 'completion_tokens'));
  let total = count(field(usage, 'total_tokens'));
  if (input !== null && output !== null) {
    const sum = input + output;
    if (!Number.isSafeInteger(sum) || (total !== null && total !== sum)) throw new Error('unsupported_usage');
    total = sum;
  }
  if (total !== null && ((input !== null && input > total) || (output !== null && output > total))) throw new Error('unsupported_usage');
  subset(usage, api === 'responses' ? 'input_tokens_details' : 'prompt_tokens_details', ['cached_tokens', 'cache_write_tokens'], input, total);
  subset(usage, api === 'responses' ? 'output_tokens_details' : 'completion_tokens_details', ['reasoning_tokens'], output, total);
  return { input_tokens: input, output_tokens: output, total_tokens: total };
}

export function extractOpenAIUsage(response, api = 'responses') {
  try { return usageSnapshot(response, api); } catch { return UNKNOWN(); }
}

function completion(response, api) {
  try {
    if (api === 'responses') {
      if (field(response, 'object') !== 'response') return { status: 'unknown', usage: UNKNOWN() };
      const status = field(response, 'status');
      return { status: status === 'completed' ? 'success' : status === 'failed' ? 'error' : 'unknown',
        usage: RESPONSE_TERMINAL.has(status) ? extractOpenAIUsage(response, api) : UNKNOWN() };
    }
    if (api === 'chat_completions') {
      if (field(response, 'object') !== 'chat.completion') return { status: 'unknown', usage: UNKNOWN() };
      const choices = array(field(response, 'choices'));
      if (choices.length !== 1 || field(choices[0], 'index') !== 0) return { status: 'unknown', usage: UNKNOWN() };
      const reason = field(choices[0], 'finish_reason');
      return { status: SUCCESS.has(reason) ? 'success' : 'unknown',
        usage: TERMINAL.has(reason) ? extractOpenAIUsage(response, api) : UNKNOWN() };
    }
  } catch { /* Invalid instrumentation metadata must not change the provider result. */ }
  return { status: 'unknown', usage: UNKNOWN() };
}

function callHandle(exporter, metadata) {
  try { return exporter.startCall(metadata); } catch { return null; }
}
function finish(call, status, usage = UNKNOWN()) {
  try { call?.finish({ status, ...usage }); } catch { /* Telemetry cannot fail the application. */ }
}
function cancelled(signal) {
  try { return signal?.aborted === true; } catch { return false; }
}

export async function openaiCall(exporter, metadata, operation, { api = 'responses', signal } = {}) {
  const call = callHandle(exporter, metadata);
  try {
    const response = await operation();
    const result = cancelled(signal) ? { status: 'unknown', usage: UNKNOWN() } : completion(response, api);
    finish(call, result.status, result.usage);
    return response;
  } catch (error) {
    finish(call, cancelled(signal) ? 'unknown' : 'error');
    throw error;
  }
}

class StreamState {
  constructor(api) {
    this.api = api;
    this.id = null;
    this.terminal = null;
    this.usage = null;
    this.badUsage = false;
    this.invalid = !APIS.has(api);
    this.error = false;
  }
  bind(value) {
    const id = field(value, 'id');
    if (typeof id !== 'string' || !/^[A-Za-z0-9_-]{1,128}(?![\s\S])/.test(id)) throw new Error('invalid_stream');
    if (this.id !== null && this.id !== id) throw new Error('invalid_stream');
    this.id = id;
  }
  mark(value) {
    if (this.terminal !== null && this.terminal !== value) throw new Error('invalid_stream');
    this.terminal = value;
  }
  snapshot(value) {
    // Retain only the freshly copied numeric values, before the caller sees a chunk.
    let next;
    try { next = usageSnapshot(value, this.api); }
    catch { this.badUsage = true; return; }
    if (this.usage !== null && Object.keys(next).some((key) => next[key] !== this.usage[key])) throw new Error('invalid_stream');
    this.usage = next;
  }
  observe(chunk) {
    if (this.invalid) return;
    try {
      if (this.api === 'responses') {
        const type = field(chunk, 'type');
        if (type === 'error') { this.error = true; return; }
        if (['response.created', 'response.in_progress'].includes(type)) {
          this.bind(field(chunk, 'response'));
          return;
        }
        if (!['response.completed', 'response.failed', 'response.incomplete'].includes(type)) return;
        const response = field(chunk, 'response');
        this.bind(response);
        const status = field(response, 'status');
        if (field(response, 'object') !== 'response' || type !== 'response.' + status) throw new Error('invalid_stream');
        this.mark(status);
        this.snapshot(response);
      } else {
        if (field(chunk, 'object') !== 'chat.completion.chunk') throw new Error('invalid_stream');
        this.bind(chunk);
        const choices = array(field(chunk, 'choices'));
        if (choices.length) {
          if (field(choices[0], 'index') !== 0) throw new Error('invalid_stream');
          const reason = field(choices[0], 'finish_reason');
          if (reason !== undefined && reason !== null) {
            if (!TERMINAL.has(reason)) throw new Error('invalid_stream');
            this.mark(reason);
          }
          // Usage attached to non-final delta chunks is not a completion snapshot.
        } else if (field(chunk, 'usage') !== undefined && field(chunk, 'usage') !== null) {
          if (this.terminal === null) throw new Error('invalid_stream');
          this.snapshot(chunk);
        }
      }
    } catch { this.invalid = true; this.usage = null; }
  }
  result() {
    if (this.invalid || (this.error && this.terminal !== null && this.terminal !== 'failed')) return { status: 'unknown', usage: UNKNOWN() };
    if (this.error) return { status: 'error', usage: this.badUsage ? UNKNOWN() : this.usage ?? UNKNOWN() };
    const status = this.api === 'responses'
      ? this.terminal === 'completed' ? 'success' : this.terminal === 'failed' ? 'error' : 'unknown'
      : SUCCESS.has(this.terminal) ? 'success' : 'unknown';
    return { status, usage: this.terminal === null || this.badUsage ? UNKNOWN() : this.usage ?? UNKNOWN() };
  }
}

export async function* openaiStream(exporter, metadata, operation, { api = 'responses', signal } = {}) {
  const call = callHandle(exporter, metadata);
  const state = new StreamState(api);
  let stream, iterator, complete = false, settled = false, yielding = false;
  try {
    stream = await operation();
    iterator = stream[Symbol.asyncIterator]?.() ?? stream[Symbol.iterator]();
    while (true) {
      const next = await iterator.next();
      if (next.done) break;
      state.observe(next.value);
      yielding = true;
      yield next.value;
      yielding = false;
    }
    complete = true;
    const result = cancelled(signal) ? { status: 'unknown', usage: UNKNOWN() } : state.result();
    settled = true;
    finish(call, result.status, result.usage);
  } catch (error) {
    settled = true;
    finish(call, yielding || cancelled(signal) ? 'unknown' : 'error');
    throw error;
  } finally {
    if (!settled) finish(call, 'unknown');
    // Early return/throw does not automatically close a manually-driven iterator.
    if (!complete) {
      try { await iterator?.return?.(); } catch { /* Preserve the original outcome. */ }
      try { stream?.controller?.abort?.(); } catch { /* OpenAI Stream exposes its request controller. */ }
    }
  }
}

File: examples/native-capture/guardian_openai.py
"""Content-free OpenAI usage adapters for an application-owned SDK operation.

No SDK import, client construction, credentials, content serialization or pricing.
Only scalar usage and lifecycle metadata survive inspection of a provider object.
"""
import asyncio
import inspect
import re
from types import GetSetDescriptorType, MemberDescriptorType

_MAX = 2**53 - 1
_MISSING = object()
_ID = re.compile(r"[A-Za-z0-9_-]{1,128}\Z")
_APIS = ("responses", "chat_completions")
_RESPONSE_TERMINAL = {"completed": "success", "failed": "error", "incomplete": "unknown"}
_CHAT_TERMINAL = {"stop": "success", "tool_calls": "success", "function_call": "success",
                  "length": "unknown", "content_filter": "unknown"}


def _unknown():
    return {"input_tokens": None, "output_tokens": None, "total_tokens": None}


def _stored(value, attribute="__dict__"):
    if type(value) is dict:
        return value
    # Read real Pydantic model storage via its builtin __dict__ descriptor. This
    # bypasses instance getters and never calls model_dump/dict/custom serializers.
    cls = type(value)
    bases = type.__getattribute__(cls, "__mro__")
    if not any(type.__getattribute__(base, "__module__") in ("pydantic.main", "pydantic.v1.main")
               and type.__getattribute__(base, "__name__") == "BaseModel" for base in bases):
        raise ValueError()
    for base in bases:
        descriptor = type.__getattribute__(base, "__dict__").get(attribute)
        if type(descriptor) in (GetSetDescriptorType, MemberDescriptorType):
            stored = descriptor.__get__(value, cls)
            if type(stored) is dict:
                return stored
            if attribute == "__pydantic_extra__" and stored is None:
                return {}
            raise ValueError()
    if attribute == "__pydantic_extra__":
        # Pydantic v1 stores extras in __dict__, without a separate slot.
        return {}
    raise ValueError()


def _get(value, name, default=_MISSING):
    stored = _stored(value)
    if name in stored:
        return dict.get(stored, name)
    if type(value) is not dict:
        # New provider fields may live in v2's extra storage. Inspect only the
        # requested allowlisted key through its builtin descriptor, never model
        # getters, serializers or the remaining provider payload.
        return dict.get(_stored(value, "__pydantic_extra__"), name, default)
    return default


def _counter(value):
    if value is _MISSING or value is None:
        return None
    if type(value) is not int or not 0 <= value <= _MAX:
        raise ValueError()
    return value


def _usage(response, api):
    if type(api) is not str or api not in _APIS:
        raise ValueError()
    usage = _get(response, "usage")
    if usage is _MISSING or usage is None:
        return _unknown()
    inputs, outputs = ("input_tokens", "output_tokens") if api == "responses" else ("prompt_tokens", "completion_tokens")
    inp, out, total = (_counter(_get(usage, name)) for name in (inputs, outputs, "total_tokens"))
    if inp is not None and out is not None:
        summed = _counter(inp + out)
        if total is not None and total != summed:
            raise ValueError()
        total = summed
    if total is not None and any(number is not None and number > total for number in (inp, out)):
        raise ValueError()
    details = (("input_tokens_details", ("cached_tokens", "cache_write_tokens"), inp), ("output_tokens_details", ("reasoning_tokens",), out)) if api == "responses" else (
        ("prompt_tokens_details", ("cached_tokens", "cache_write_tokens"), inp), ("completion_tokens_details", ("reasoning_tokens",), out))
    for field, children, parent in details:
        container = _get(usage, field)
        if container is _MISSING or container is None:
            continue
        for child in children:
            subset = _counter(_get(container, child))
            if subset is not None and any(limit is not None and subset > limit for limit in (parent, total)):
                raise ValueError()
    return {"input_tokens": inp, "output_tokens": out, "total_tokens": total}


def extract_openai_usage(response, api="responses"):
    """Return only validated numeric usage; malformed snapshots become unknown."""
    try:
        return _usage(response, api)
    except BaseException:
        return _unknown()


def _chat_reason(response):
    choices = _get(response, "choices")
    if type(choices) is not list or len(choices) != 1:
        return None
    choice = choices[0]
    index, reason = _get(choice, "index"), _get(choice, "finish_reason")
    if type(index) is not int or index != 0 or type(reason) is not str or reason not in _CHAT_TERMINAL:
        return None
    return reason


def _completion(response, api):
    if type(api) is not str or api not in _APIS:
        return "unknown", _unknown()
    expected = "response" if api == "responses" else "chat.completion"
    kind = _get(response, "object")
    if type(kind) is not str or kind != expected:
        return "unknown", _unknown()
    if api == "responses":
        terminal = _get(response, "status")
        if type(terminal) is str and terminal in _RESPONSE_TERMINAL:
            return _RESPONSE_TERMINAL[terminal], extract_openai_usage(response, api)
    elif api == "chat_completions":
        reason = _chat_reason(response)
        if reason is not None:
            return _CHAT_TERMINAL[reason], extract_openai_usage(response, api)
    return "unknown", _unknown()


def _start(exporter, metadata):
    try:
        return exporter.start_call(**metadata)
    except BaseException:
        return None


def _finish(handle, status, usage=None):
    try:
        if handle is not None:
            handle.finish(status, **(usage or _unknown()))
    except BaseException:
        pass


def _result(handle, response, api):
    try:
        status, usage = _completion(response, api)
    except BaseException:
        status, usage = "unknown", _unknown()
    _finish(handle, status, usage)


def _error_status(error):
    return "error" if isinstance(error, Exception) else "unknown"


def openai_call(exporter, operation, *, api="responses", agent_name, model, trace_id=None, parent_observation_id=None):
    handle = _start(exporter, dict(agent_name=agent_name, model=model, trace_id=trace_id,
                                  parent_observation_id=parent_observation_id))
    try:
        result = operation()
    except BaseException as error:
        _finish(handle, _error_status(error))
        raise
    _result(handle, result, api)
    return result


async def async_openai_call(exporter, operation, *, api="responses", agent_name, model, trace_id=None, parent_observation_id=None):
    handle = _start(exporter, dict(agent_name=agent_name, model=model, trace_id=trace_id,
                                  parent_observation_id=parent_observation_id))
    try:
        result = operation()
        if inspect.isawaitable(result):
            result = await result
    except BaseException as error:
        _finish(handle, _error_status(error))
        raise
    _result(handle, result, api)
    return result


class _StreamState:
    """One bounded ID, marker and numeric tuple; never retain provider chunks."""
    def __init__(self, api):
        self.api = api
        self.identifier = self.terminal = self.usage = None
        self.valid_usage = None
        self.bad_usage = self.error = False
        self.conflict = type(api) is not str or api not in _APIS

    def _bind(self, value):
        if type(value) is not str or not _ID.fullmatch(value):
            self.conflict = True
        elif self.identifier is None:
            self.identifier = value
        elif self.identifier != value:
            self.conflict = True

    def _terminal(self, marker):
        if self.terminal is not None and self.terminal != marker:
            self.conflict = True
        self.terminal = marker

    def _usage(self, response):
        try:
            usage = tuple(_usage(response, self.api).values())
        except BaseException:
            self.bad_usage = True
            self.usage = (None, None, None)
            return
        if self.valid_usage is not None and self.valid_usage != usage:
            self.conflict = True
        self.valid_usage = usage
        if not self.bad_usage:
            self.usage = usage

    def observe(self, chunk):
        try:
            self._observe(chunk)
        except BaseException:
            self.conflict = True

    def _observe(self, chunk):
        if self.conflict:
            return
        if self.api == "responses":
            event_type = _get(chunk, "type")
            if type(event_type) is not str:
                return
            if event_type == "error":
                self.error = True
                return
            if event_type not in (
                    "response.created", "response.in_progress",
                    "response.completed", "response.failed", "response.incomplete"):
                return
            response = _get(chunk, "response")
            self._bind(_get(response, "id"))
            terminal = event_type.removeprefix("response.")
            if terminal in _RESPONSE_TERMINAL:
                status, kind = _get(response, "status"), _get(response, "object")
                if type(status) is not str or status != terminal or type(kind) is not str or kind != "response":
                    self.conflict = True
                self._terminal(terminal)
                self._usage(response)
        else:
            kind = _get(chunk, "object")
            if type(kind) is not str or kind != "chat.completion.chunk":
                self.conflict = True
                return
            self._bind(_get(chunk, "id"))
            choices = _get(chunk, "choices")
            if type(choices) is not list or len(choices) > 1:
                self.conflict = True
                return
            usage = _get(chunk, "usage")
            if not choices:
                if usage is not _MISSING and usage is not None:
                    if self.terminal is None:
                        self.conflict = True
                    self._usage(chunk)
                return
            # Usage on choice/delta chunks is not a final completion snapshot.
            choice = choices[0]
            if type(_get(choice, "index")) is not int or _get(choice, "index") != 0:
                self.conflict = True
                return
            reason = _get(choice, "finish_reason")
            if reason is not _MISSING and reason is not None:
                if type(reason) is not str or reason not in _CHAT_TERMINAL:
                    self.conflict = True
                else:
                    self._terminal(reason)

    def result(self):
        if self.conflict or self.error and self.terminal not in (None, "failed"):
            return "unknown", _unknown()
        if self.error:
            return "error", dict(zip(("input_tokens", "output_tokens", "total_tokens"), self.usage or (None, None, None)))
        if self.terminal is None:
            return "unknown", _unknown()
        if self.api == "responses":
            status = _RESPONSE_TERMINAL.get(self.terminal, "unknown")
        else:
            status = _CHAT_TERMINAL.get(self.terminal, "unknown")
        return status, dict(zip(("input_tokens", "output_tokens", "total_tokens"), self.usage or (None, None, None)))


def _close(stream, iterator):
    for value in (iterator, stream if stream is not iterator else None):
        try:
            if value is not None:
                closer = getattr(value, "close", None)
                if closer is not None:
                    closer()
        except BaseException:
            pass


async def _aclose(stream, iterator):
    for value in (iterator, stream if stream is not iterator else None):
        try:
            if value is not None:
                closer = getattr(value, "aclose", None) or getattr(value, "close", None)
                if closer is not None:
                    result = closer()
                    if inspect.isawaitable(result):
                        await result
        except BaseException:
            pass


def openai_stream(exporter, operation, *, api="responses", agent_name, model, trace_id=None, parent_observation_id=None):
    handle = _start(exporter, dict(agent_name=agent_name, model=model, trace_id=trace_id,
                                  parent_observation_id=parent_observation_id))
    state, stream, iterator = _StreamState(api), None, None
    status, usage = "unknown", _unknown()
    yielding = False
    try:
        stream = operation()
        iterator = iter(stream)
        while True:
            try:
                chunk = next(iterator)
            except StopIteration as completed:
                status, usage = state.result()
                return completed.value
            state.observe(chunk)
            yielding = True
            yield chunk
            yielding = False
            del chunk
    except BaseException as error:
        status, usage = "unknown" if yielding else _error_status(error), _unknown()
        raise
    finally:
        _finish(handle, status, usage)
        _close(stream, iterator)


async def async_openai_stream(exporter, operation, *, api="responses", agent_name, model, trace_id=None, parent_observation_id=None):
    handle = _start(exporter, dict(agent_name=agent_name, model=model, trace_id=trace_id,
                                  parent_observation_id=parent_observation_id))
    state, stream, iterator = _StreamState(api), None, None
    status, usage = "unknown", _unknown()
    yielding = False
    try:
        stream = operation()
        if inspect.isawaitable(stream):
            stream = await stream
        iterator = aiter(stream)
        while True:
            try:
                chunk = await anext(iterator)
            except StopAsyncIteration:
                status, usage = state.result()
                return
            state.observe(chunk)
            yielding = True
            yield chunk
            yielding = False
            del chunk
    except BaseException as error:
        status, usage = "unknown" if yielding else _error_status(error), _unknown()
        raise
    finally:
        _finish(handle, status, usage)
        await _aclose(stream, iterator)
